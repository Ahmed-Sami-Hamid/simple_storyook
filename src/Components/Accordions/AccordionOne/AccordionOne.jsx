// components
import Spacer from "../../../Components/UI-Kit/Spacer/Spacer";

// libraries
import { useState } from "react";
import PropTypes from "prop-types";

// icons
import { BiChevronDown } from "react-icons/bi";

// styling
import Style from "./AccordionOne.module.css";

const AccordionOne = ({
  indicator = {
    icon: <BiChevronDown />,
    activeClassName: Style.activeIcon,
  },
  heading = "Heading Lorem ipsum dolor sit amet, consectetur adipiscing elit heading",
  content = "Content Lorem ipsum dolor sit amet consectetur adipisicing elit. Dicta nisi tempora consectetur autem minima exercitationem quibusdam, similique esse deleniti maiores!",
  children,
}) => {
  const [isActive, setIsActive] = useState(false);

  return (
    <div className={Style.accordion}>
      <div
        className={Style.head}
        onClick={() => setIsActive(!isActive)}
        data-testid="accordion-head"
      >
        <h3 className={Style.heading}>{heading}</h3>
        <div
          className={!isActive ? Style.icon : indicator.activeClassName}
          data-testid="accordion-icon"
        >
          {indicator.icon}
        </div>
      </div>
      <Spacer sm />
      <div
        className={!isActive ? Style.body : Style.activeBody}
        data-testid="accordion-body"
      >
        {children || <p className={Style.content}>{content}</p>}
      </div>
    </div>
  );
};

// prop-types checking
const { string, object } = PropTypes;

AccordionOne.propTypes = {
  indicator: object,
  heading: string.isRequired,
};

export default AccordionOne;
