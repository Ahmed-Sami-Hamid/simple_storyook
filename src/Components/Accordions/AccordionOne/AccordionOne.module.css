/* fonts */
@import url("https://fonts.googleapis.com/css2?family=Poppins:ital,wght@0,400;0,500;0,600;0,700;0,800;0,900;1,400;1,500;1,600;1,700;1,800;1,900&display=swap");

* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

/* accordion */
.accordion {
  overflow: hidden;
}

/* head */
.head {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 20px 50px;
  gap: 20px;
  background-color: #fafafa;
  border-radius: 50px;
  cursor: pointer;
}

/* heading */
.heading {
  font-size: 1.5rem;
  font-weight: 500;
  font-family: "Poppins", sans-serif;
  color: #060c3c;
}

/* icon */
.icon > svg,
.icon > img,
.activeIcon > svg,
.activeIcon > img {
  width: 30px;
  height: 30px;
  transition: all 0.5s ease;
}

.activeIcon > svg,
.activeIcon > img {
  transform: rotate(-180deg);
}

/* body */
.body,
.activeBody {
  position: relative;
  padding: 20px 50px;
  z-index: -1;
  background-color: #ffffff;
  transform-origin: top;
  transform: translateY(-200%);
  transition: transform 0.7s ease;
}

.activeBody {
  transform: unset;
}

/* content */
.content {
  font-size: 1.3rem;
  font-family: "Poppins", sans-serif;
  font-weight: 300;
  color: #5d6180;
}

@media (max-width: 480px) {
  .head {
    flex-direction: column;
  }
  .heading,
  .content {
    text-align: center;
  }
}
