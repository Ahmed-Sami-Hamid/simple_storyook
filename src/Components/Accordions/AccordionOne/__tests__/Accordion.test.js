// components
import AccordionOne from "../AccordionOne";

// enzyme
import Enzyme, { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";

// @testing-library/react
import { render, fireEvent, screen } from "@testing-library/react";

// snapshot test
Enzyme.configure({ adapter: new Adapter() });

describe("snapshot testing of component render", () => {
  it("should render with no problem", () => {
    const wrapper = shallow(<AccordionOne heading="Test Heading" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

// unit test
describe("Head click effect on state and icon", () => {
  it("should change isActive state and in turn change of the icon className upon clicking the head element", () => {
    render(<AccordionOne heading="Test Heading" />);
    expect(screen.getByTestId("accordion-icon").className).toBe("icon");
    fireEvent.click(screen.getByTestId("accordion-head"));
    expect(screen.getByTestId("accordion-icon").className).toBe("activeIcon");
  });
});

describe("Props checking", () => {
  describe("Indicator property", () => {
    it("should use the className provided in activeClassName property", () => {
      render(
        <AccordionOne
          heading="Test Heading"
          indicator={{ activeClassName: "testActiveClass" }}
        />
      );
      fireEvent.click(screen.getByTestId("accordion-icon"));
      expect(screen.getByTestId("accordion-icon").className).toBe(
        "testActiveClass"
      );
    });
    it("should use the icon provided in icon property", () => {
      render(
        <AccordionOne
          heading="Test Heading"
          indicator={{
            icon: <img alt="this is provided icon" />,
          }}
        />
      );
      expect(screen.getByTestId("accordion-icon").firstChild.alt).toBe(
        "this is provided icon"
      );
    });
    it("should not render the content if the chidlren is provided", () => {
      render(
        <AccordionOne heading="Test Heading" content="content text">
          <h2>chidlren</h2>
        </AccordionOne>
      );
      expect(
        screen.getByTestId("accordion-body").firstChild.textContent
      ).toEqual("chidlren");
    });
  });
});
