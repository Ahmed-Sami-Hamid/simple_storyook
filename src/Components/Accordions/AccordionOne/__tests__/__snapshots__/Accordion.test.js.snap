// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing of component render should render with no problem 1`] = `
<div
  className="accordion"
>
  <div
    className="head"
    data-testid="accordion-head"
    onClick={[Function]}
  >
    <h3
      className="heading"
    >
      Test Heading
    </h3>
    <div
      className="icon"
      data-testid="accordion-icon"
    >
      <BiChevronDown />
    </div>
  </div>
  <Spacer
    sm={true}
  />
  <div
    className="body"
    data-testid="accordion-body"
  >
    <p
      className="content"
    >
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Dicta nisi tempora consectetur autem minima exercitationem quibusdam, similique esse deleniti maiores!
    </p>
  </div>
</div>
`;
